notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b1e4a2e1-2fbb-420b-a052-5e80885e79c1",
+   "metadata": {},
+   "source": [
+    "## SQL Based Ingestion\n",
+    "\n",
+    "Let's start with something simple. With this first example we are loading all the data from an external file into the table called \"wikipedia_events\"."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fd47fcff-9055-4058-852f-6f5c61d07965",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql = '''\n",
+    "REPLACE INTO \"wikipedia_events\" OVERWRITE ALL\n",
+    "WITH \"ext\" AS \n",
+    "(\n",
+    "    SELECT *\n",
+    "    FROM TABLE(\n",
+    "      EXTERN(\n",
+    "        '{\"type\":\"http\",\"uris\":[\"https://druid.apache.org/data/wikipedia.json.gz\"]}',\n",
+    "        '{\"type\":\"json\"}'\n",
+    "      )\n",
+    ") EXTEND (\"isRobot\" VARCHAR, \"channel\" VARCHAR, \"timestamp\" VARCHAR, \"flags\" VARCHAR, \"isUnpatrolled\" VARCHAR, \"page\" VARCHAR, \"diffUrl\" VARCHAR, \"added\" BIGINT, \"comment\" VARCHAR, \"commentLength\" BIGINT, \"isNew\" VARCHAR, \"isMinor\" VARCHAR, \"delta\" BIGINT, \"isAnonymous\" VARCHAR, \"user\" VARCHAR, \"deltaBucket\" BIGINT, \"deleted\" BIGINT, \"namespace\" VARCHAR, \"cityName\" VARCHAR, \"countryName\" VARCHAR, \"regionIsoCode\" VARCHAR, \"metroCode\" BIGINT, \"countryIsoCode\" VARCHAR, \"regionName\" VARCHAR))\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"timestamp\") AS \"__time\",\n",
+    "  *\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "druid.display.run_task(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1776cd84-9f38-40a3-9e47-5e821d15c901",
+   "metadata": {},
+   "source": [
+    "`REPLACE` or `INSERT` at the beginning of the statement tells Druid to execute an ingestion task. `INSERT` is used when appending data, `REPLACE` when replacing data. Both methods work when adding data to a new or empty Druid datasource. The `OVERWRITE ALL` clause means that the whole datasource will be replaced with the result of this ingestion. \n",
+    "\n",
+    "```\n",
+    "REPLACE INTO \"wikipedia_events\" OVERWRITE ALL\n",
+    "```\n",
+    "\n",
+    "The `WITH` clause is used to declare one or more input sources, this could also be placed directly in the `FROM` clause of the final `SELECT`, but this is easier to read:\n",
+    "\n",
+    "```\n",
+    "WITH \"ext\" AS \n",
+    "(\n",
+    "    SELECT ...\n",
+    "      FROM TABLE (EXTERN ( ... ) )\n",
+    ") EXTEND (...)\n",
+    "```\n",
+    "\n",
+    "`EXTERN` supports many batch [input sources](https://druid.apache.org/docs/latest/ingestion/native-batch-input-sources.html) and [formats](https://druid.apache.org/docs/latest/ingestion/data-formats.html). In this case we are using input source type `http` to access a set or `uris` that each contain a data file in the `json` data format. Note that compressed files are allowed and will automatically be decompressed.\n",
```

Comment:
... In this case the SQL statement specified an input source type 'http' ...

(I tend to avoid you / we)

Reply:
done